mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,175 @@
+    fn sp_blocking_write(port: *mut c_void, buf: *const u8, count: usize, timeout_ms: u32) -> c_int;
+}
+
+#[derive(Debug, Clone, Copy)]
+struct ControllerState {
+    buttons: u8,
+    adc_value: u16,
+}
+
+fn open_port(port_name: &str) -> Result<*mut c_void, String> {
+    let port_name_c = CString::new(port_name).expect("CString::new failed");
+    let mut port: *mut c_void = ptr::null_mut();
+
+    unsafe {
+        if sp_get_port_by_name(port_name_c.as_ptr(), &mut port) != 0 {
+            return Err(format!("Failed to get port by name {}", port_name));
+        }
+        if sp_open(port, 3) != 0 {
+            sp_free_port(port);
+            return Err(format!("Failed to open port {}", port_name));
+        }
+    }
+    Ok(port)
+}
+
+fn send_command(port: *mut c_void, command: Command) -> Result<(), String> {
+    let bytes = command.as_bytes();
+    unsafe {
+        let written = sp_blocking_write(port, bytes.as_ptr(), bytes.len(), 1000);
+        if written != bytes.len() as i32 {
+            return Err(format!(
+                "{:?} failed to send {:?}: {}",
+                port, command, written
+            ));
+        }
+    }
+    Ok(())
+}
+
+fn read_controller_state(port: *mut c_void) -> Result<ControllerState, String> {
+    let mut buffer = [0u8; 3];
+    unsafe {
+        let bytes_read = sp_blocking_read(port, buffer.as_mut_ptr(), buffer.len(), 1000);
+        if bytes_read != buffer.len() as i32 {
+            return Err(format!(
+                "{:?} failed to read controller state: {}",
+                port, bytes_read
+            ));
+        }
+    }
+    let adc_value = u16::from_be_bytes([buffer[1], buffer[2]]);
+    let buttons = buffer[0];
+    // println!("Read buttons: {:08b}, ADC value: {}", buttons, adc_value); // Debug print
+    Ok(ControllerState {
+        buttons,
+        adc_value,
+    })
+}
+
+pub fn controller(
+    port_name: &str,
+    receiver: Receiver<(String, String)>,
+    model_sender: Sender<(String, String)>,
+    player: usize,
+) {
+    let port = match open_port(port_name) {
+        Ok(port) => port,
+        Err(e) => {
+            eprintln!("{}", e);
+            return;
+        }
+    };
+    model_sender
+        .send((player.to_string(), "".to_string()))
+        .unwrap();
+
+    let mut last_button_state = 0;
+    let mut last_debounce_time = Instant::now();
+    let debounce_delay = std::time::Duration::from_millis(50);
+
+    loop {
+        // Wait for game to start
+        if let Ok((_, message)) = receiver.recv() {
+            if message == "Escape" {
+                break;
+            }
+        }
+
+        // Start LED command sequence
+        loop {
+            if let Ok((_, message)) = receiver.recv() {
+                if let Err(e) = send_command(port, Command::from_string(&message).unwrap()) {
+                    eprintln!("Failed to send command: {}", e);
+                };
+                if message == "Start" {
```

Comment:
also, I think I made this comment earlier, but having these messages be explicitly enums would make this a lot cleaner, and more robust - you could directly match on the received message to ensure all expected paths are accounted for, would also make adding new commands easier